### DISCUSSION TITLE
Exposición «Algoritmos libres de bloqueo»: Preguntas y respuestas

### DISCUSSION THREAD

**Discussion by gwolf** [unamfi/sistop-2022-1#180] [Q&A]
Abro este foro como un espacio de discusión para la exposición que prepararon los alumnos Marcos Bryan Díaz Hernández y Nestor Enrique Fuerte Martínez, «Algoritmos libres de bloqueo». El video está disponible en:

https://www.youtube.com/watch?v=3SdDXVJMUfg

Pueden consultar también el documento de texto en que los alumnos presentan este mismo tema:

https://github.com/unamfi/sistop-2022-1/blob/main/exposiciones/D%C3%ADazBryan-FuerteNestor/Algoritmos%20Libre%20de%20Bloqueo%20-%20Documento%20escrito.pdf

Además de participar aquí, les pido que llenen la pequña encuesta que está disponible en:

https://www.questionpro.com/t/ARiwuZqIrF

**Comment by JesusESD** [comment]
¿Qué tal compañeros? Los saludo con mucho gusto. El tema de los bloqueos siempre me ha parecido muy interesante, por esa razón les pregunto ¿Cuál de las fuentes que aparecen en sus referencias recomendarían para seguir aprendiendo acerca de estas cuestiones? Al final de su video casi no se hizo mención de la bibliografía, así que se los pregunto por aquí. Les agradeceré mucho su respuesta. :)

**Reply by BrainDead59** [reply]
Hola, gran parte de la investigación esta basada en la serie de artículos publicados en  LWN: https://lwn.net/Kernel/Index/#Lockless_algorithms, si también te interesan las implementaciones en las estructuras de datos te paso también el url: https://www.scss.tcd.ie/jeremy.jones/CS4021/lockless.pdf

**Comment by CarlosGMendozaB** [comment]
Hola compañeros, me pareció muy buena su presentación. Y con respecto al tema es muy interesante el funcionamiento de estos dichos algoritmos. ¿Conocen o saben donde se implementan estos algoritmos en otros campos a parte de lo que se vio en clase y lo explicado en la presentación?. Y otra pregunta, ¿Los procesos o hilos pueden apoyar a otro proceso o hilo para que termine ? Gracias.

**Reply by BrainDead59** [reply]
Respondiendo la primer pregunta estos algoritmos sirven para poder manejar situaciones donde se presentan bloqueos, ahora si lo llevamos a una aplicación real, sirve para poder modelar situaciones reales donde existen bloqueos, hay ejemplos que modelan ejemplos de aplicaciones con ciertas condiciones: https://storage.googleapis.com/coding-competitions.appspot.com/HC/2018/hashcode2018_qualification_task.pdf como en este problema de una competencia que es meramente teórico. También hay más contenido teórico en el siguiente url: https://www.baeldung.com/lock-free-programming

Con respecto a la segunda pregunta, un proceso es ejecutado mediante los multihilos lo que facilita el procesamiento, pero dos procesos no pueden ayudarse directamente, debido a que si ambos procesos manejan la misma información para hacer más rápida la ejecución nos podríamos encontrar un caso de bloqueo mutuo. Sin embargo los hilos pueden comunicarse para saber en que momento es necesaria su ejecución dentro de un proceso mediante señales, lo que permite sincronizar la ejecución de los hilos. Igual te comparto un documento para que lo puedas checar: https://web.cs.ucdavis.edu/~matloff/matloff/public_html/158/PLN/OrigS06/ParProc.pdf

**Comment by Inakyord** [comment]
Hola, espero estén bien. Se me hizo interesante el tema, pero no me quedó del todo claro dónde se utilizan estos algoritmos. ¿Se implementan por el programador dentro de su código multihilo, ó son parte de la administración de procesos del sistema operativo? Otra duda que me gustaría explicarán es ¿en qué consiste el problema ABA que mencionan al final del escrito y por qué CAS falla? Gracias.

**Reply by BrainDead59** [reply]
Respondiendo la primer pregunta estos algoritmos se ven implementados por los desarrolladores del kernel, debido a la concurrencia entre los distintos hilos al momento de acceder a memoria. En cuanto a la segunda el problema ABA esta relacionado con CAS debido a la forma en que funciona CAS. 
CAS implementa verificaciones del valor de una variable que quiera modificar, el primer parámetro es la localidad de memoria de la variable, el segundo es el ultimo valor que se leyó de la variable y el tercero es el valor que se le desea asignar a la variable. Lo que se hace es comprobar si el ultimo valor leído es el mismo que existe en memoria, si son iguales entonces no hubo modificación por parte de otro proceso, por lo que se permite que se haga la modificación.
Sin embargo aquí entra el problema ABA ya que puedes tener un hilo haciendo modificaciones a la memoria y que este sea interrumpido cuando se encuentra dentro del if, y ahora entra un segundo proceso que modifica la información que estaba trabajando el hilo 1. Al regresar el proceso 1 ya no hace la verificación sino continua con su ejecución pero con información modificada lo que genera un comportamiento impredecible. 
Es posible ver el problema ABA en las estructuras de datos libres del bloqueo, igual te comparto un url donde lo explican mas a detalle:
https://www.baeldung.com/cs/aba-concurrency
